### PR TITLE
Add support for interrupting child threads created by RunThreaded.

### PR DESCRIPTION
**Pull request by skschneider**
Addresses #916.

Pressing CTRL+C after resource-provisioning threads have been created
will now cause a KeyboardInterrupt to be raised in each child thread.

- Move RunThreaded, RunParallelThreads, and RunParallelProcesses out of
  vm_util.py into their own file.
- Combine the bodies of RunParallelThreads and RunParallelProcesses
  into a single _RunParallelTasks function that uses separate task
  managers for each.
- Raise KeyboardInterrupt explicitly in each child thread.

### CODE REVIEW

**Comment by skschneider** [comment]
@ehankland PTAL

File: perfkitbenchmarker/background_tasks.py [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+# long timeouts is used instead. This is because some of Python's built-in wait
+# methods are non-interruptable without a timeout.
+_LONG_TIMEOUT = 1000.
+
+# Constants used for polling waits. See _WaitForCondition.
+_WAIT_MIN_RECHECK_DELAY = 0.001  # 1 ms
+_WAIT_MAX_RECHECK_DELAY = 0.050  # 50 ms
+
+
+def _GetCallString(target_arg_tuple):
```
Comment (ehankland) [inline perfkitbenchmarker/background_tasks.py]: Looks like you haven't deleted the copy of this function in vm_util - I couldn't find any references to the old function - do you still need it or can it be deleted?

Comment (skschneider) [inline perfkitbenchmarker/background_tasks.py]: Deleted.

File: perfkitbenchmarker/background_tasks.py [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+
+  def __init__(self):
+    self._deque = deque()
+    self._lock = threading.Lock()
+    self._lock.acquire()
+
+  def _WaitForItem(self):
+    self._lock.acquire()
+
+  def _SignalAvailableItem(self):
```
Comment (ehankland) [inline perfkitbenchmarker/background_tasks.py]: Seems like [condition variables](https://docs.python.org/2/library/threading.html#condition-objects) might be appropriate here.

File: perfkitbenchmarker/vm_util.py [hunk 3]
```diff
@@ -194,243 +188,9 @@ def GetSshOptions(ssh_key_filename):
...
-        msg = ('Exception occurred while calling {0}:{1}{2}'.format(
-            _GetCallString(target_arg_tuples[index]), os.linesep, stacktrace))
-        logging.error(msg)
-        error_strings.append(msg)
-  if error_strings:
-    msg = ('The following exceptions occurred during parallel execution:'
-           '{0}{1}'.format(os.linesep, os.linesep.join(error_strings)))
-    raise errors.VmUtil.CalledProcessException(msg)
-  return results
+RunParallelProcesses = background_tasks.RunParallelProcesses
```
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Doesn't look like there are any references to vm_util.RunParallelProcesses (or background_tasks.RunParallelProcesses for that matter). Who/what do we expect to use this?

File: perfkitbenchmarker/vm_util.py [hunk 4]
```diff
@@ -194,243 +188,9 @@ def GetSshOptions(ssh_key_filename):
...
+RunParallelThreads = background_tasks.RunParallelThreads
```
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Looks like there are only 3 references to vm_util.RunParallelThreads (and those are all in benchmark_spec.py). Maybe just make those use background_tasks.RunParallelThreads instead and get rid of this line.

File: perfkitbenchmarker/vm_util.py [hunk 5]
```diff
@@ -194,243 +188,9 @@ def GetSshOptions(ssh_key_filename):
...
+RunThreaded = background_tasks.RunThreaded
```
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Is the expectation that we are eventually going to change all RunThreaded calls to point to background_tasks.RunThreaded or are we keeping this indefinitely? Maybe add a TODO if the former.

File: perfkitbenchmarker/background_tasks.py (hunk 2 above)
Comment (skschneider) [inline perfkitbenchmarker/background_tasks.py]: Unfortunately not. They're not resilient to KeyboardInterrupts. Internally, they use two threading.Lock objects. One of them is the "waiter" Lock used to wait for the notification, similar to how this _lock is used to wait for items. But in order to start waiting or to send the notification, the current thread needs ownership of the other Lock (let's call it the "condition" Lock). That's where things can get tricky. Because the KeyboardInterrupt can be raised after the condition Lock is acquired, but before it's released, things can end up in a state where that condition Lock is forever tied up.

File: perfkitbenchmarker/vm_util.py (hunk 4 above)
Comment (skschneider) [inline perfkitbenchmarker/vm_util.py]: Added TODO. Will update references and remove vm_util.RunParallelThreads in a separate PR.

File: perfkitbenchmarker/vm_util.py (hunk 5 above)
Comment (skschneider) [inline perfkitbenchmarker/vm_util.py]: This can be a separate PR to redirect references, but we should keep the vm_util version at least until the v1.5.0 release.

File: perfkitbenchmarker/background_tasks.py (hunk 2 above)
Comment (skschneider) [inline perfkitbenchmarker/background_tasks.py]: Here's something you can try for yourself. Write this to a file, run it, and hit CTRL+C. Repeat a few times and see if any bad behaviors show up. Then try it using an RLock instead of a Lock.
```
import itertools
import threading

c = threading.Condition(threading.Lock())
try:
  # Infinite loop of just acquiring and releasing a lock.
  for _ in itertools.repeat([None]):
    with c:
      pass
except KeyboardInterrupt:
  print 'KeyboardInterrupt. If cleaned up properly, expect c to have been released.'
  try:
    # Release again to confirm. If error, c was already released. If no error,
    # c was not released the first time.
    c.release()
    print 'But c had not been released.'
  except threading.ThreadError:
    print 'And c had indeed been released.'
```

Comment (ehankland) [inline perfkitbenchmarker/background_tasks.py]: Thanks for the script - I saw the bad behavior - that is definitely weird. Was there a problem using a condition variable with a normal (non-reentrant) lock?

Comment (skschneider) [inline perfkitbenchmarker/background_tasks.py]: Yes, the Condition class has this problem whether it uses a Lock or RLock. To simplify the with statement, it appears to be something like
```
__enter__()
try:
  ...
finally:
  __exit__(...)
```
And the gap between enter and try can still be interrupted.

File: perfkitbenchmarker/background_tasks.py [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+    self._task_queues[worker_id].Put((task_id, task))
+
+  def AwaitAnyTask(self):
+    worker_id, task_id = self._response_queue.Get()
+    self._available_worker_ids.append(worker_id)
+    return task_id
+
+  def HandleKeyboardInterrupt(self):
+    # Raise a KeyboardInterrupt in each child thread.
+    for thread in self._threads:
```
Comment (ehankland) [inline perfkitbenchmarker/background_tasks.py]: Do we want all threads to log the keyboard interrupt? Depending on the number of threads this can be a lot of output

Comment (skschneider) [inline perfkitbenchmarker/background_tasks.py]: It could be helpful to know where they were interrupted. But at the same time, threads that are not actively processing a task will likely be interrupted while within _ExecuteBackgroundThreadTasks, so that could be a lot of useless information. The only way that I can think of to differentiate helpful from unhelpful cases is to inspect the stack of the exception, which I would want to be a separate thread. How about if I switch this to debug level for now (but still include the exc_info) and make a separate PR to reduce output?

Comment (ehankland) [inline perfkitbenchmarker/background_tasks.py]: SGTM

**Comment by ehankland** [comment]
LGTM

File: perfkitbenchmarker/background_tasks.py (hunk 6 above)
Comment (skschneider) [inline perfkitbenchmarker/background_tasks.py]: Oops, did I say "thread"? I meant PR. Anyway, I'll go ahead with the log level change.

Comment (skschneider) [inline perfkitbenchmarker/background_tasks.py]: Changed to debug and added a TODO.

**Comment by skschneider** [comment]
GitHub has been a bit weird and unresponsive today. Merged manually (https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/commit/00d68725aa94edffabe4632c9547463bb6ddef7b), so closing this PR.
